smolyay/basis.py

```diff
@@ -429,3 +480,65 @@ def __init__(self, basis_functions):
     def basis_functions(self):
         """list: Basis functions."""
         return self._basis_functions
+
+    def __len__(self):
+        return len(self.basis_functions)
+
+    def __getitem__(self, key):
+        return self.basis_functions[key]
+
+
+class NestedBasisFunctionSet(BasisFunctionSet):
+    """Set of nested basis functions and sample points.
+
+    Parameters
+    ----------
+    basis_functions : list
+        Basis functions in set.
+
+    num_per_level : list
+        number of unique functions per level
+
+    Raises
+    ------
+    IndexError
+        number of basis function does not match functions in each level
+    """
+
+    def __init__(self, basis_functions, num_per_level):
```

Comment:
I think a better constructor for this would be to take only `basis_functions` as a list of lists. You can then compute the number per level, etc. from this, and flatten it to pass it to the parent.
```suggestion
    def __init__(self, basis_functions):
```

Reply:
That would actually be a harder to generate than what is currently used. With the current method, it's easy to get the `num_per_level` by having a `NestedUnidimensionalPointSet` generate it for you. Then `basis_functions` can be generated the same way as the unnested version.

Requiring `basis_functions` to be list of lists would be more cumbersome because there's nothing in the code that generates that data structure, so the user would have to do it themselves.

We could create `GrowthMethod` classes that specify a specific method of generating `num_per_level`. Then each `NestedUnidimensionalPointSet` hardcodes what GrowthMethod it uses, and `NestedBasisFunctionSet` takes a GrowthMethod as a parameter. That allows us to avoid making new nested basis classes while reusing the code from the point sets to generate `num_per_level`.